### DISCUSSION TITLE
python-musicbrainzngs needs a replacement

### DISCUSSION THREAD

**Discussion by Louson** [beetbox/beets#4660] [Ideas]
As explained in #4651 , the musicbrainz python library [python-musicbrainzngs](https://github.com/alastair/python-musicbrainzngs) receives no update.

It is used in beets auto-tagging and in several plugins.

As @sampsyo truly says, it was created when the only output format was xml and has a lot of parsing code that can be deleted. Also, after digging in it, it reveals some lack of conception, using sometimes oriented object programming and sometimes global variables.

It can indeed be replaced by a much simpler library. Of course, this implies to find who and when. But first, two question have to be answered:

## Q1. Implement the library inside this project or as an independent project

Both have some pros and cons.

The cons about implementing it inside beets is that it will add some maintainance work to the beets developpers.
The cons about an independant project is that it requires at least one developer who actively maintains it, with the risk for it to be abandoned again.

I think at this stage it is better to implement into beets project unless we know some others projects that use this and who will appreciate to share the effort.

##  Q2. JSON abstraction or simple pass-through

We have two options:

* the library is a simple pass-through and it is up to the caller to parse the json
* the library has an interface that hides the JSON. The advantage is we don't have to change the caller when the API changes. It gives more stability over time

**Comment by sampsyo** [comment]
I'd be open to any and all opinions on this general direction, but here are two gut reactions:

* There is an "if you build it, they will come" aspect to this kind of library. It could make a lot of sense to start the project as an independent library under the Beetbox org. It seems like many projects would at least potentially be interested in using it, and the more (non-beets) users there are, the more we can leverage development effort to build a high-quality library we can all use.
* A simple pass-through of the JSON responses seems very much the right thing to me. This would avoid the need to make changes to the library to adapt when new stuff is added to the API. (This is avoids the trap that the current library is in: new features have been added to the API, but adding support for them in the library requires boring data-munging work to expose the new data.) And breaking changes to the MB API are rare and well-publicized, so it makes sense to externalize this risk to end applications.

**Reply by RollingStar** [reply]
> A simple pass-through of the JSON responses seems very much the right thing to me. This would avoid the need to make changes to the library to adapt when new stuff is added to the API. (This is avoids the trap that the current library is in: new features have been added to the API, but adding support for them in the library requires boring data-munging work to expose the new data.) And breaking changes to the MB API are rare and well-publicized, so it makes sense to externalize this risk to end applications.

How would this look? I agree in principle that it should be something like `response['album']` rather than making a bespoke Album class that has to be updated with every MB update. But I wonder if even a light-touch passthrough would be better off in its own library, to let people use it outside of beets. There's MB auth, rate limits, and the possibility of querying your own local MB server to consider.

**Reply by sampsyo** [reply]
Oh yes! I completely agree with this, if I didn't make that clear:

> even a light-touch passthrough would be better off in its own library, to let people use it outside of beets

I would love to develop this as a separate library that lives on PyPI. Publishing the library (even if it's a simple pass-through thing) would really help leverage the support of the community.

**Reply by Louson** [reply]
> I would love to develop this as a separate library that lives on PyPI. Publishing the library (even if it's a simple pass-through thing) would really help leverage the support of the community.

I am not a python expert. I won't be able to help you on packaging and maintaining. But I can help you on the development. So far I have inspected the musicbrainzngs. I have started something based on it, including:
- dropping xml parsing stuff
- changing the authentication process, including oauth2
- setting https as default
- changing the way the format (json or xml) is defined
- changing tests to match https, json and authentication

**Reply by Serene-Arc** [reply]
I might be interested in rewriting the library. It seems like a fun project.

**Reply by Serene-Arc** [reply]
Ah nevermind! I didn't notice the date for this.

**Comment by Louson** [comment]
There is an extra question about credentials (password or oauth). The former library used to store the login/password. In my opinion, the library should not do that and the credentials should only be managed by the caller. What is your opinion ?

**Reply by sampsyo** [reply]
Right; a clean-slate implementation would probably want to be OAuth-only (or at least OAuth-first).

**Comment by Louson** [comment]
Hello, here is a suggestion https://gitlab.com/Louson/python-musicbrainzez/-/merge_requests/1

It looks huge but most of it is duplicating xml tests in json.

Can you have a look ?
It lacks packaging and documentation fix, and probably some organization.

I kept the parsing in a dedicated subfolder, but it is not hold by default.
The JSON API is requested by default, but the xml API can be requested in any request.
I renamed the library. I am not sure this is relevant but this actually gives the possibility to create a musicbrainzngs/musicbrainz.py that adds backward compatibility.
I disabled the password storage and added oauth2

**Reply by RollingStar** [reply]
Did a semi thorough skim of the non test parts. Skimmed the tests. Thank you for this!

Good "port" of the functionality which makes it easier to review.

Once this review is done, you may want to make more changes that make the code easier to read/maintain.

User agent should? be stored in a consts.py or similar, then "import USERAGENT from consts"

Would it make sense to abbreviate the import?

import musicbrainzez as mez

I tried finding best practice. It looks like pandas imports "pandas", but it's a lot easier to write than "musicbrainzez"

Should this change?
    classifiers=[
        "Development Status :: 5 - Production/Stable",

For readers, make_barcode_request and the funcs around it are removed because they come back in another file.

It would have been easier to review this if the test data was in its own separate PR submitted after the core PR was done.

Same with having tests in their own PR.

**Reply by Louson** [reply]
Hello, thanks for the review.

Actually, every commit is very singular and would require its own PR.
It is more relevant to check it commit by commit skipping examples and tests.

This is not an official PR, but only a suggestion for someone who can carry the package.

**Reply by Louson** [reply]
Hello @sampsyo, would you like to also give it a view ?

**Comment by Louson** [comment]
Hello, I might have some time for this. Do you have any idea how to go further ?

**Reply by sampsyo** [reply]
Hi! That's cool! If I were getting this started, I think I would begin with a Python project depending only on the `requests` library. The first order of business would be to create some kind of "API" class you can instantiate with an optional endpoint URL, etc., and then add a few methods for looking up objects by ID. For example, maybe a method `get_release(id)` would essentially just fetch a URL like this:
https://musicbrainz.org/ws/2/release-group/6f46d5e4-1e46-4480-b120-d04b37145f7f?fmt=json

(Or, more appropriately, the base URL with an appropriate `Accept` header.) It would then just parse that JSON response and return it. The interesting stuff comes when adding error handling, rate limiting, etc.…

**Reply by Louson** [reply]
Hello, do you want to have a look on this ?

https://gitlab.com/Louson/python-mbzero

It handles requests and oauth2. Any advice on python styling is good

**Reply by sampsyo** [reply]
Wow; looks cool at first glance! Any chance there is some documentation ready yet? That would be helpful for knowing how beets might go about using it.

**Reply by Louson** [reply]
If you except authenticatio, the best documentation so far is musicbrainz API description (https://wiki.musicbrainz.org/MusicBrainz_API). It's only making lookup/browse/search.

I think my goal will be to match to this API as good as possible. While Musicbrainzngs was an abstraction layer, this won't be. That will make this lib easy to maintain. It might mean more work for clients. Looking at beets, I believe there is no big effort.

First of all, every call such as  `if "work-level-rels" in musicbrainzngs.VALID_BROWSE_INCLUDES['recording']:` shall be dropped. There is too much effort to maintain a library that double checks every call is correct. Clients will have to check the Musicbrainz documentation or receive an exception that matches an HTTP error.

Second, there won't be any function such as `browse_recordings`. The client will have to call the lookup/browse/search with right arguments.

The main question that remains is about contents. Who handles xml/json formatting and parsing ? Perhaps the lib can handle dictionaries to json (and reverse) and also leave the ability for the client to provide xml content. I don't know if musicbrainz submissions accept json content. As beets does not make submissions, that will come later.

**Reply by arsaboo** [reply]
I checked our beets code and looks like we are only using couple of calls - `get_work_by_id`, `search_recordings`, `search_releases`. As long as these work, we should be able to transition to `mbzero`. Any bugfixes can be easily made in the mbzero library.

**Reply by sampsyo** [reply]
Maybe it would be interesting to try a prototype… that is, to see what a fork of beets would look like to use this library. Perhaps running the tests would uncover if anything major is missing.

**Reply by Louson** [reply]
Yes. I think I first need the support of the digest authentication. I left it unimplemented

**Reply by Louson** [reply]
I'm thinking about a first step which would isolate the musicbrainzngs usage in beets (in a file mb_interface.py). Run the tests to check, and then replace the dep to musicbrainzngs. Perhaps files in beetsplug can also use it.

So far I have troubles in running tests. It doesn't pass on my distrib, and neither on a fresh debian buster (I replaced py38 by py37 in tox.ini). It looks to miss dependencies to sphinx and mypy. Also, py37-mypy fails. Is that only me ?

**Reply by sampsyo** [reply]
> I'm thinking about a first step which would isolate the musicbrainzngs usage in beets (in a file mb_interface.py). Run the tests to check, and then replace the dep to musicbrainzngs. Perhaps files in beetsplug can also use it.

For the most part, this is already the case: `mb.py` is the central place where we interact with MusicBrainz, I believe. There may be some exceptions in plugins, but for a proof of concept, it's probably OK to leave them depending on the old library.

> So far I have troubles in running tests. It doesn't pass on my distrib, and neither on a fresh debian buster (I replaced py38 by py37 in tox.ini). It looks to miss dependencies to sphinx and mypy. Also, py37-mypy fails. Is that only me ?

Got it! Sphinx is hard to get right in Tox because it's an `sdist`-time dependency to build the manpage. We are probably missing a dependency on mypy. However, yes, `mypy` is expected to fail—you can ignore that for now.

**Reply by Louson** [reply]
Please have a check on PR #4936

**Reply by RollingStar** [reply]
How does the PR relate to https://gitlab.com/Louson/python-musicbrainzez/-/merge_requests/1 ? I can't view this anymore , if there's unique code in there can you upload a zip or something so someone else can continue it? Thanks.

**Reply by Louson** [reply]
Hello, I archived the project as I made more interest to the rework https://gitlab.com/mbzero/python-mbzero after @sampsyo 's suggestions. Unfortunately #4936 does not apply because of the massive lint done after my push. I had a look and it didn't look easy to update the PR. I must work on it, but perhaps it's possible to have some review before I spend time in it ? Thanks

**Reply by RollingStar** [reply]
That's okay, I was just looking for a newer Python API for Musicbrainz.

**Reply by Louson** [reply]
Would you like to help ? There is probably a lot of things to do, mainly tests before getting it stable

**Reply by RollingStar** [reply]
Not in the next few months, sorry.

**Reply by Louson** [reply]
Hello, I have updated the repository https://gitlab.com/mbzero/python-mbzero
Please make feedbacks if you try it

**Comment by Louson** [comment]
To resume the current situation,

We have a PR #5976 which is suggestion for a replacement of musicbrainzngs with mbzero
There is a new PR #6052 which aims to drop musicbrainzngs in profit of direct http requests

It seems to me this new PR has a simpler point of view and perhaps it's a better choice when comes the question of maintaining the code.
On the other side, mbzero offers a complete interface to the musicbrainz API, including OAuth2 authentication, which was the initial goal. It might be a better option to uniform connections to musicbrainz API across plugins and other projects. However, there's currently no multiple software using musicbrainzngs (I know only whipper so far)

Both seems fine for me if we manage to support OAuth2 if at least we move from musicbrainzngs which is 4 years unmaintained.

Perhaps @nicomem will have a better overview

**Reply by nicomem** [reply]
You resumed the situation quite well.

I took a quick look at the new `request`-based PR and it seems to be quite similar to the `mbzero`-based with the obvious difference of removing a layer of abstraction by directly making the requests.

Apart from the difference of making requests directly vs with an external lib, it should end up in quite similar code when all plugins will be migrated to the new solution (for example, the new PR would probably need to use a shared instance to be usable by the other plugins -> https://github.com/beetbox/beets/pull/6052#discussion_r2417750241).

In the end, this new PR has its pros&cons:
- By removing a layer of abstraction done in an external lib, this allows to specialize for beets and forgo some genericity that would be required in library code, making the code simpler
- It also simplifies the job of beets developers as changes can simply and quickly be done depending on beets needs
- However, it does not allow other (and potentially future) projects to benefit from the work done as it is not directly usable as a library and may not be easily extracted to one as it may be/become too specialized to beets needs

All in all, I'm also fine with both solutions, so long as we are able to move from the long-unmaintained `musicbrainzngs` lib.
